rpki-validator/src/main/java/net/ripe/rpki/validator3/util/HappyEyeballsResolver.java

```diff
@@ -273,52 +270,55 @@ private static Runnable dnsLookupRunnable(Name hostname, int queryType, Queue<Op
-                if (addressOptional != null) {
-                    if (addressOptional.isPresent()) {
-                        final InetAddress address = addressOptional.get();
-                        try {
-                            log.debug("Connecting to {}", address);
-                            final SocketChannel socketChannel = SocketChannel.open();
-                            socketChannel.configureBlocking(false);
-                            socketChannel.connect(new InetSocketAddress(address, port));
-                            socketChannels.add(Optional.of(socketChannel));
-                        } catch (IOException e) {
-                            // can't connect - move on
-                            log.debug("Error connecting to {}", address);
-                        } finally {
-                            TimeUnit.MILLISECONDS.sleep(CONNECTION_ATTEMPT_DELAY_MS);
-                        }
-                    } else {
-                        return false;
-                    }
+            final boolean haveMoreAF2 = openConnectionFromQueue(secondAfAddresses, port, selector);
+            connectedSocket = awaitSuccessfulConnection(selector, CONNECTION_ATTEMPT_DELAY_MILLIS);
+            if (connectedSocket != null) return connectedSocket;
+
+            // watch out! do not inline these booleans,
+            // otherwise you'll shortcircuit second evaluation
+            keepGoing = (haveMoreAF1 || haveMoreAF2) && haveTime(deadline);
+        }
+        if (selector.keys().isEmpty()) {
+            return Optional.empty();
+        } else {
+            final long remainingTime = Math.max(0, deadline - System.nanoTime());
+            return awaitSuccessfulConnection(selector, TimeUnit.NANOSECONDS.toMillis(remainingTime));
+        }
+    }
+
+    private static boolean openConnectionFromQueue(Queue<Optional<InetAddress>> queue,
+                                                   int port,
+                                                   Selector selector) {
+        final Optional<InetAddress> addressOptional = queue.poll();
+        if (addressOptional != null) {
+            if (addressOptional.isPresent()) {
+                final InetAddress address = addressOptional.get();
+                try {
+                    log.info("Connecting to {}", address);
+                    final SocketChannel socketChannel = SocketChannel.open();
+                    socketChannel.configureBlocking(false);
+                    socketChannel.register(selector, SelectionKey.OP_CONNECT);
+                    socketChannel.connect(new InetSocketAddress(address, port));
+                } catch (IOException e) {
+                    // can't connect - move on
+                    log.info("Error connecting to {}", address);
                 }
-                return true;
+            } else {
+                return false;
             }
-        };
+        }
+        return true;
```

Comment:
Curious how null could be inserted to the queue, and why we returned true in the open connection from queue in those cases.

Reply:
Well, javadoc literally says " Retrieves and removes the head of this queue, or returns {@code null} if this queue is empty", so it means the queue is empty.

Reply:
For a moment I thought the poll itself is returning Optional (present/empty) but it's the entry type on the queue that was optional. Clear now. Still shouldn't it return false when there is nothing on the queue?

Reply:
It returns either the element or "nothing", "null" is just java's way to say "nothing" :)

Reply:
The empty queue does not mean the end of the queue (or that we went through all addresses).
The signal of the end of the queue is Optional.empty in the queue. This is when we return false (meaning there's nothing more to wait for in this address family). In all other cases we return true.